### ISSUE TITLE
Pitch extraction for FastSpeech2

### ISSUE THREAD

**Issue by iamanigeeit** [espnet/espnet#4545]
According to the FastSpeech2 paper page 14,

> 1) We use linear interpolation to fill the unvoiced frame in pitch contour
> 2) We transform the resulting pitch contour to logarithmic scale
> 3) We normalize it to zero mean and unit variance for each utterance, and we have to save the original utterance-level mean and variance for pitch contour reconstruction
> 4) We convert the normalized pitch contour to pitch spectrogram using continuous wavelet transform

For (1) and (2), i can see that the `Dio` pitch extractor does it correctly.

For (3), i need to set `pitch_normalize` in [tts.sh](https://github.com/espnet/espnet/blob/master/egs2/TEMPLATE/tts1/tts.sh).
- Why is pitch_normalize and energy_normalize set to none in `tts.sh` line 659? It is ignoring the yaml config.
- Does the normalizer save the original mean and variance?

For (4), i can't find the wavelet transform anywhere. I have no idea how to implement this because it we enforce `pitch length == duration length == text length` to align the input timesteps. But the transform relies on equally spaced wavelet positions, whereas the durations are variable!

> take the pitch spectrogram as the training target for the pitch predictor which is optimized with MSE loss.

Currently, we are doing MSE loss on `p_outs` and `ps`, which are the log F0 pitch values, not spectrogram.

In the paper, page 4,

> To take the pitch contour as input in both training and inference, we quantize pitch F0 (ground-truth/predicted value for train/inference respectively) of each frame to 256 possible values in log-scale and further convert it into pitch embedding vector p and add it to the expanded hidden sequence.

My understanding is: take the log-F0 values, quantize into `{0, 1, ... 255}` and convert into a one-hot vector, then put it through an embedding layer. This seems different from the `conv1d` we are using.

If i train without (3) and (4), i get unstable pitch loss:

![pitch_loss](https://user-images.githubusercontent.com/13131208/181693759-cb04608a-6095-4b2a-8ee5-674b2a44c49e.png)
![loss](https://user-images.githubusercontent.com/13131208/181693764-a42f4704-8470-4284-9e90-18bfa3e335ef.png)

**Comment by kan-bayashi** [comment]
> Why is pitch_normalize and energy_normalize set to none in tts.sh line 659? It is ignoring the yaml config.
> Does the normalizer save the original mean and variance?

In Stage 5, we calculate mean and variance of pitch and energy for using normalizer. Therefore, we do not need to use normalizer in Stage 5.

**Comment by kan-bayashi** [comment]
For question (4), we use FastPitch-style strategy rather than FastSpeech2.
https://github.com/espnet/espnet/blob/92ea573f44e252644860b6b7906916218e4ab78b/espnet2/tts/fastspeech2/fastspeech2.py#L36-L44
The following discussion may help you.
https://github.com/espnet/espnet/issues/2019#issuecomment-657169925

**Comment by iamanigeeit** [comment]
> For question (4), we use FastPitch-style strategy rather than FastSpeech2.
> 
> https://github.com/espnet/espnet/blob/92ea573f44e252644860b6b7906916218e4ab78b/espnet2/tts/fastspeech2/fastspeech2.py#L36-L44
> 
> 
> The following discussion may help you.
> [#2019 (comment)](https://github.com/espnet/espnet/issues/2019#issuecomment-657169925)

This is a very helpful discussion, thank you.

Do you have any suggestion on the unstable pitch loss?

> > Why is pitch_normalize and energy_normalize set to none in tts.sh line 659? It is ignoring the yaml config.
> > Does the normalizer save the original mean and variance?
> 
> In Stage 5, we calculate mean and variance of pitch and energy for using normalizer. Therefore, we do not need to use normalizer in Stage 5.

Just to confirm what i understand from reading the code and stats output... After stage 5, the output in data/stats/logdir are not normalized. The normalization is done during the `forward` call on the `ESPnetTTSModel`.

**Comment by kan-bayashi** [comment]
> Just to confirm what i understand from reading the code and stats output... After stage 5, the output in data/stats/logdir are not normalized. The normalization is done during the forward call on the ESPnetTTSModel.

Right. Normalization is done in `forward`.
https://github.com/espnet/espnet/blob/4a0a9338eeb3d778d9dbb23bae199d122c242f75/espnet2/tts/espnet_model.py#L93-L125

> Do you have any suggestion on the unstable pitch loss?

It should depend on the training data. Is your training data like a narration?
If the data is like a story-telling, the prediction will be much difficult.
And tensorboard shows iter-based loss for the training.
How about the epoch-based loss ? (which should be shown in images/ folder)

**Comment by iamanigeeit** [comment]
You are right, i am using LJSpeech and the pitch loss looks OK after longer training. I can't see the epoch loss well because the X are too close together.
